third_party/gfootball_engine/src/systems/graphics/rendering/opengl_renderer3d.cpp

```diff
@@ -1688,6 +1716,15 @@ void OpenGLRenderer3D::UpdateTexture(int textureID, SDL_Surface *source,
   int w = source->w;
   int h = source->h;
 
+  SDL_LockSurface(source);
+  mapping.glTexSubImage2D(GL_TEXTURE_2D, 0, x, y, w, h, type2, GL_UNSIGNED_BYTE,
+          source->pixels);
+  SDL_UnlockSurface(source);
+
+  if (mipmaps) {
+    DO_VALIDATION;
+    mapping.glGenerateMipmap(GL_TEXTURE_2D);
```

Comment:
Same again, mipmaps should be generated after `glTexSubImage2D` call: 
https://www.khronos.org/opengl/wiki/Common_Mistakes#Automatic_mipmap_generation